File: WPF/Controls/PopupHolder.vb
Imports System.Windows.Markup
Imports Ks.Common.MVVM

Namespace Controls

    <ContentProperty("Content")>
    Public Class PopupHolder
        Inherits FrameworkElement ' UIElement is not used as base type throughout the whole libraries! Using FrameworkElement instead.

        Protected Overrides Function MeasureOverride(availableSize As Size) As Size
            Return Nothing
        End Function

        Protected Overrides Function ArrangeOverride(finalSize As Size) As Size
            Return Nothing
        End Function

        Protected Overridable Function ArrangePopup(ByVal Panel As PopupPanel, ByVal PanelSize As Size, ByVal PopupSize As Size) As Rect?
            Dim Target = Me.GetTarget()

            If Target Is Nothing Then
                Return Nothing
            End If

            Dim Transform = Target.TransformToVisual(Panel)
            Dim TargetSize = New Size(Target.ActualWidth, Target.ActualHeight)
            Dim BasePoint = Transform.Transform(New Point(TargetSize.Width * Me.TargetHorizontalAnchor, TargetSize.Height * Me.TargetVerticalAnchor))

            BasePoint -= New Vector(PopupSize.Width * Me.PopupHorizontalAnchor, PopupSize.Height * Me.PopupVerticalAnchor)

            Return New Rect(BasePoint, PopupSize)
        End Function

#Region "ShowPopup Command"
        Private _ShowPopupCommand As DelegateCommand = New DelegateCommand(AddressOf Me.ShowPopup)

        Public Sub ShowPopup()
            Me.IsPopupShown = True
        End Sub

        Public ReadOnly Property ShowPopupCommand As DelegateCommand
            Get
                Return Me._ShowPopupCommand
            End Get
        End Property
#End Region

#Region "HidePopup Command"
        Private _HidePopupCommand As DelegateCommand = New DelegateCommand(AddressOf Me.HidePopup)

        Public Sub HidePopup()
            Me.IsPopupShown = False
        End Sub

        Public ReadOnly Property HidePopupCommand As DelegateCommand
            Get
                Return Me._HidePopupCommand
            End Get
        End Property
#End Region

#Region "Content Property"
        Public Shared ReadOnly ContentProperty As DependencyProperty = DependencyProperty.Register("Content", GetType(UIElement), GetType(PopupHolder), New PropertyMetadata(Nothing, AddressOf Content_Changed))

        Private Shared Sub Content_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, PopupHolder)

            'Dim OldValue = DirectCast(E.OldValue, UIElement)
            Dim NewValue = DirectCast(E.NewValue, UIElement)
            Dim Popup = TryCast(NewValue, Popup)

            If Popup IsNot Nothing Then
                Self.Popup = Popup
            Else
                If Self.Popup IsNot Nothing Then
                    Self.Popup.Content = NewValue
                End If
            End If
        End Sub

        Public Property Content As UIElement
            Get
                Return DirectCast(Me.GetValue(ContentProperty), UIElement)
            End Get
            Set(ByVal value As UIElement)
                Me.SetValue(ContentProperty, value)
            End Set
        End Property
#End Region

#Region "Target Property"
        Public Shared ReadOnly TargetProperty As DependencyProperty = DependencyProperty.Register("Target", GetType(FrameworkElement), GetType(PopupHolder), New PropertyMetadata(Nothing, AddressOf Target_Changed))

        Private Shared Sub Target_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, PopupHolder)

            'Dim OldValue = DirectCast(E.OldValue, FrameworkElement)
            'Dim NewValue = DirectCast(E.NewValue, FrameworkElement)
        End Sub

        Public Property Target As FrameworkElement
            Get
                Return DirectCast(Me.GetValue(TargetProperty), FrameworkElement)
            End Get
            Set(ByVal value As FrameworkElement)
                Me.SetValue(TargetProperty, value)
            End Set
        End Property

        Private Function GetTarget() As FrameworkElement
            Return If(Me.Target, TryCast(Media.VisualTreeHelper.GetParent(Me), FrameworkElement))
        End Function
#End Region

#Region "Popup Property"
        Private Shared ReadOnly PopupPropertyKey As DependencyPropertyKey = DependencyProperty.RegisterReadOnly("Popup", GetType(Popup), GetType(PopupHolder), New PropertyMetadata(Nothing, AddressOf Popup_Changed))

        Private Shared Sub Popup_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, PopupHolder)

            Dim OldValue = DirectCast(E.OldValue, Popup)
            Dim NewValue = DirectCast(E.NewValue, Popup)

            If OldValue IsNot Nothing Then
                OldValue.ArrangeCallBack = Nothing
            End If
            If NewValue IsNot Nothing Then
                NewValue.ArrangeCallBack = AddressOf Self.ArrangePopup
            End If
        End Sub

        Public Shared ReadOnly PopupProperty As DependencyProperty = PopupPropertyKey.DependencyProperty

        Public Property Popup As Popup
            Get
                Return DirectCast(Me.GetValue(PopupProperty), Popup)
            End Get
            Private Set(ByVal value As Popup)
                Me.SetValue(PopupPropertyKey, value)
            End Set
        End Property
#End Region

#Region "IsPopupShown Property"
        Public Shared ReadOnly IsPopupShownProperty As DependencyProperty = DependencyProperty.Register("IsPopupShown", GetType(Boolean), GetType(PopupHolder), New PropertyMetadata(False, AddressOf IsPopupShown_Changed))

        Private Shared Sub IsPopupShown_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, PopupHolder)

            'Dim OldValue = DirectCast(E.OldValue, Boolean)
            Dim NewValue = DirectCast(E.NewValue, Boolean)

            Dim Popup = Self.Popup
            If Popup Is Nothing And NewValue Then
                Popup = New Popup() With {.Content = Self.Content}
                Self.Popup = Popup
            End If

            If Popup IsNot Nothing Then
                'If NewValue Then
                '    Self.OnBeforeShowPopup()
                'End If
                Popup.IsShown = NewValue
            End If
        End Sub

        Public Property IsPopupShown As Boolean
            Get
                Return DirectCast(Me.GetValue(IsPopupShownProperty), Boolean)
            End Get
            Set(ByVal value As Boolean)
                Me.SetValue(IsPopupShownProperty, value)
            End Set
        End Property
#End Region

#Region "TargetHorizontalAnchor Property"
        Public Shared ReadOnly TargetHorizontalAnchorProperty As DependencyProperty = DependencyProperty.Register("TargetHorizontalAnchor", GetType(Double), GetType(PopupHolder), New PropertyMetadata(0.5, AddressOf TargetHorizontalAnchor_Changed))

        Private Shared Sub TargetHorizontalAnchor_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, PopupHolder)

            Dim OldValue = DirectCast(E.OldValue, Double)
            Dim NewValue = DirectCast(E.NewValue, Double)
        End Sub

        Public Property TargetHorizontalAnchor As Double
            Get
                Return DirectCast(Me.GetValue(TargetHorizontalAnchorProperty), Double)
            End Get
            Set(ByVal value As Double)
                Me.SetValue(TargetHorizontalAnchorProperty, value)
            End Set
        End Property
#End Region

#Region "TargetVerticalAnchor Property"
        Public Shared ReadOnly TargetVerticalAnchorProperty As DependencyProperty = DependencyProperty.Register("TargetVerticalAnchor", GetType(Double), GetType(PopupHolder), New PropertyMetadata(0.5, AddressOf TargetVerticalAnchor_Changed))

        Private Shared Sub TargetVerticalAnchor_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, PopupHolder)

            Dim OldValue = DirectCast(E.OldValue, Double)
            Dim NewValue = DirectCast(E.NewValue, Double)
        End Sub

        Public Property TargetVerticalAnchor As Double
            Get
                Return DirectCast(Me.GetValue(TargetVerticalAnchorProperty), Double)
            End Get
            Set(ByVal value As Double)
                Me.SetValue(TargetVerticalAnchorProperty, value)
            End Set
        End Property
#End Region

#Region "PopupHorizontalAnchor Property"
        Public Shared ReadOnly PopupHorizontalAnchorProperty As DependencyProperty = DependencyProperty.Register("PopupHorizontalAnchor", GetType(Double), GetType(PopupHolder), New PropertyMetadata(0.5, AddressOf PopupHorizontalAnchor_Changed))

        Private Shared Sub PopupHorizontalAnchor_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, PopupHolder)

            Dim OldValue = DirectCast(E.OldValue, Double)
            Dim NewValue = DirectCast(E.NewValue, Double)
        End Sub

        Public Property PopupHorizontalAnchor As Double
            Get
                Return DirectCast(Me.GetValue(PopupHorizontalAnchorProperty), Double)
            End Get
            Set(ByVal value As Double)
                Me.SetValue(PopupHorizontalAnchorProperty, value)
            End Set
        End Property
#End Region

#Region "PopupVerticalAnchor Property"
        Public Shared ReadOnly PopupVerticalAnchorProperty As DependencyProperty = DependencyProperty.Register("PopupVerticalAnchor", GetType(Double), GetType(PopupHolder), New PropertyMetadata(0.5, AddressOf PopupVerticalAnchor_Changed))

        Private Shared Sub PopupVerticalAnchor_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, PopupHolder)

            Dim OldValue = DirectCast(E.OldValue, Double)
            Dim NewValue = DirectCast(E.NewValue, Double)
        End Sub

        Public Property PopupVerticalAnchor As Double
            Get
                Return DirectCast(Me.GetValue(PopupVerticalAnchorProperty), Double)
            End Get
            Set(ByVal value As Double)
                Me.SetValue(PopupVerticalAnchorProperty, value)
            End Set
        End Property
#End Region

    End Class

End Namespace

File: WPF/Controls/PopupHolder.cs
using System.Windows;
using System.Windows.Markup;

using Ks.Common.MVVM;

namespace Ks.Common.Controls
{
    [ContentProperty("Content")]
    public class PopupHolder : FrameworkElement // UIElement is not used as base type throughout the whole libraries! Using FrameworkElement instead.
    {

        public PopupHolder()
        {
            this.ShowPopupCommand = new DelegateCommand(this.ShowPopup);
            this.HidePopupCommand = new DelegateCommand(this.HidePopup);
        }

        protected override Size MeasureOverride(Size availableSize)
        {
            return default;
        }

        protected override Size ArrangeOverride(Size finalSize)
        {
            return default;
        }

        protected virtual Rect? ArrangePopup(PopupPanel Panel, Size PanelSize, Size PopupSize)
        {
            var Target = this.GetTarget();

            if (Target == null)
            {
                return default;
            }

            var Transform = Target.TransformToVisual(Panel);
            var TargetSize = new Size(Target.ActualWidth, Target.ActualHeight);
            var BasePoint = Transform.Transform(new Point(TargetSize.Width * this.TargetHorizontalAnchor, TargetSize.Height * this.TargetVerticalAnchor));

            BasePoint -= new Vector(PopupSize.Width * this.PopupHorizontalAnchor, PopupSize.Height * this.PopupVerticalAnchor);

            return new Rect(BasePoint, PopupSize);
        }

        public void ShowPopup()
        {
            this.IsPopupShown = true;
        }

        public DelegateCommand ShowPopupCommand { get; }

        public void HidePopup()
        {
            this.IsPopupShown = false;
        }

        public DelegateCommand HidePopupCommand { get; }

        public static readonly DependencyProperty ContentProperty = DependencyProperty.Register("Content", typeof(UIElement), typeof(PopupHolder), new PropertyMetadata(null, Content_Changed));

        private static void Content_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var Self = (PopupHolder) D;

            // Dim OldValue = DirectCast(E.OldValue, UIElement)
            var NewValue = (UIElement) E.NewValue;
            var Popup = NewValue as Popup;

            if (Popup != null)
            {
                Self.Popup = Popup;
            }
            else if (Self.Popup != null)
            {
                Self.Popup.Content = NewValue;
            }
        }

        public UIElement Content
        {
            get => (UIElement) this.GetValue(ContentProperty);
            set => this.SetValue(ContentProperty, value);
        }

        public static readonly DependencyProperty TargetProperty = DependencyProperty.Register("Target", typeof(FrameworkElement), typeof(PopupHolder), new PropertyMetadata(null, Target_Changed));

        private static void Target_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var Self = (PopupHolder) D;
        }

        public FrameworkElement Target
        {
            get => (FrameworkElement) this.GetValue(TargetProperty);
            set => this.SetValue(TargetProperty, value);
        }

        private FrameworkElement GetTarget()
        {
            return this.Target ?? (System.Windows.Media.VisualTreeHelper.GetParent(this) as FrameworkElement);
        }

        private static readonly DependencyPropertyKey PopupPropertyKey = DependencyProperty.RegisterReadOnly("Popup", typeof(Popup), typeof(PopupHolder), new PropertyMetadata(null, Popup_Changed));

        private static void Popup_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var Self = (PopupHolder) D;

            var OldValue = (Popup) E.OldValue;
            var NewValue = (Popup) E.NewValue;

            if (OldValue != null)
            {
                OldValue.ArrangeCallBack = null;
            }

            if (NewValue != null)
            {
                NewValue.ArrangeCallBack = Self.ArrangePopup;
            }
        }

        public static readonly DependencyProperty PopupProperty = PopupPropertyKey.DependencyProperty;

        public Popup Popup
        {
            get => (Popup) this.GetValue(PopupProperty);
            private set => this.SetValue(PopupPropertyKey, value);
        }

        public static readonly DependencyProperty IsPopupShownProperty = DependencyProperty.Register("IsPopupShown", typeof(bool), typeof(PopupHolder), new PropertyMetadata(false, IsPopupShown_Changed));

        private static void IsPopupShown_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var Self = (PopupHolder) D;

            // Dim OldValue = DirectCast(E.OldValue, Boolean)
            var NewValue = (bool) E.NewValue;

            var Popup = Self.Popup;
            if (Popup == null & NewValue)
            {
                Popup = new Popup() { Content = Self.Content };
                Self.Popup = Popup;
            }

            if (Popup != null)
            {
                // If NewValue Then
                // Self.OnBeforeShowPopup()
                // End If
                Popup.IsShown = NewValue;
            }
        }

        public bool IsPopupShown
        {
            get => (bool) this.GetValue(IsPopupShownProperty);
            set => this.SetValue(IsPopupShownProperty, value);
        }

        public static readonly DependencyProperty TargetHorizontalAnchorProperty = DependencyProperty.Register("TargetHorizontalAnchor", typeof(double), typeof(PopupHolder), new PropertyMetadata(0.5, TargetHorizontalAnchor_Changed));

        private static void TargetHorizontalAnchor_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var Self = (PopupHolder) D;

            var OldValue = (double) E.OldValue;
            var NewValue = (double) E.NewValue;
        }

        public double TargetHorizontalAnchor
        {
            get => (double) this.GetValue(TargetHorizontalAnchorProperty);
            set => this.SetValue(TargetHorizontalAnchorProperty, value);
        }

        public static readonly DependencyProperty TargetVerticalAnchorProperty = DependencyProperty.Register("TargetVerticalAnchor", typeof(double), typeof(PopupHolder), new PropertyMetadata(0.5, TargetVerticalAnchor_Changed));

        private static void TargetVerticalAnchor_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var Self = (PopupHolder) D;

            var OldValue = (double) E.OldValue;
            var NewValue = (double) E.NewValue;
        }

        public double TargetVerticalAnchor
        {
            get => (double) this.GetValue(TargetVerticalAnchorProperty);
            set => this.SetValue(TargetVerticalAnchorProperty, value);
        }

        public static readonly DependencyProperty PopupHorizontalAnchorProperty = DependencyProperty.Register("PopupHorizontalAnchor", typeof(double), typeof(PopupHolder), new PropertyMetadata(0.5, PopupHorizontalAnchor_Changed));

        private static void PopupHorizontalAnchor_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var Self = (PopupHolder) D;

            var OldValue = (double) E.OldValue;
            var NewValue = (double) E.NewValue;
        }

        public double PopupHorizontalAnchor
        {
            get => (double) this.GetValue(PopupHorizontalAnchorProperty);
            set => this.SetValue(PopupHorizontalAnchorProperty, value);
        }

        public static readonly DependencyProperty PopupVerticalAnchorProperty = DependencyProperty.Register("PopupVerticalAnchor", typeof(double), typeof(PopupHolder), new PropertyMetadata(0.5, PopupVerticalAnchor_Changed));

        private static void PopupVerticalAnchor_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var Self = (PopupHolder) D;

            var OldValue = (double) E.OldValue;
            var NewValue = (double) E.NewValue;
        }

        public double PopupVerticalAnchor
        {
            get => (double) this.GetValue(PopupVerticalAnchorProperty);
            set => this.SetValue(PopupVerticalAnchorProperty, value);
        }
    }
}
